### ISSUE TITLE
OfflineAuth slim arm + other issues

### ISSUE THREAD

**Issue by kotmatross28729** [Roadhog360/SimpleSkinBackport#12]
#### The original repository was deleted for some unknown reason, but I still have a fork with my ridiculous fix (it works for some reason)

 **Even in its current state this mod is able to work with OfflineAuth very well, but there are some minor things:**

---

## All skins have a slim hand format

This is most likely what is indicated in readme: `Alternate skin server mods should just work, but would likely need manual patching for slim arm support.`

![2024-12-08_19 33 25](https://github.com/user-attachments/assets/d44c4f3a-cca1-46b9-a95a-58d740645950)

*On skins that have a normal hand format, you can see that the hand texture moves by 1 pixel*
![2024-12-08_19 35 18](https://github.com/user-attachments/assets/ef269257-a062-4cb9-ad60-7c8714913c15)

**Comment by Roadhog360** [comment]
In OfflineAuth's AuthLib, they are returning slim arm data for every skin. This is tjeor bug not mine.  While manual patcing would indeed be needed for slim arm support, it is not the default state, meaning OfflineAuth is for some reason returning true for all skins to have slim arms. I will inspect the source code of the mod in the future to find out why.

SimpleSkinBackport is also not designed to work with extra skin data, as such, 64x128 will not be supported, since the only reason why this would exist is custom skin data.

**Comment by kotmatross28729** [comment]
> SimpleSkinBackport is also not designed to work with extra skin data, as such, 64x128 will not be supported, since the only reason why this would exist is custom skin data.

I mean the opposite, standard 64x64 does NOT work while 64x128 (not supported) works

As I wrote: "Skins < 64x128 aren't displayed correctly", not "Skins > 64x64 aren't displayed correctly"

**Comment by Roadhog360** [comment]
Oh... Weird. It might be overriding my ImageBufferDownloadPlayerSkin class with buggy code. I'll take a look at that.

**Comment by Roadhog360** [comment]
Argh I really should stay off GitHub when waking up, helps my reading comprehension

For the 64x128 skin, have you stretched the skin to be 128 pixels tall or is it a 64,64 image with an additional 64 pixels of empty space below it?

Edit: You are most likely correct.
It seems that the skin rendering issues are caused by OA cropping all skins to 64x32 when the model renderer asks for a 64x64 image, causing it to be stretched when the model renderer asks for a 64,64 image.
This would override my mixin where I make the skin images 64x64 instead.

**Comment by kotmatross28729** [comment]
64x64 stretched to 128 height

**Comment by Roadhog360** [comment]
I believe the class you showed me is overriding my code here:

https://github.com/Roadhog360/SimpleSkinBackport/blob/master/src%2Fmain%2Fjava%2Froadhog360%2Fsimpleskinbackport%2Fclient%2FImageBufferDownloadPlayerSkin.java

**Comment by kotmatross28729** [comment]
Yeah, most likely this is so. 
Perhaps you can simply: Late mixin - Overwrite/Inject?

**Comment by Roadhog360** [comment]
I may submit a PR in the future. I wonder even what their class is doing; what does OfflineAuth do to skins to need this?

**Comment by kotmatross28729** [comment]
As far as I remember, it will take a skin with a height of 64, and convert it with a height of 32

**Comment by Roadhog360** [comment]
That is exactly what it is doing, and seems to be the root cause of the first issue.

**Comment by kotmatross28729** [comment]
I've done some testing and I don't see a pattern

![Test111](https://github.com/user-attachments/assets/9779ce66-ffb3-45f2-9c01-fb72db1bc313)



I can explain why height 64 doesn't work [1)](https://github.com/kotmatross28729/OfflineAuth/blob/1e48b6e8fc445ac28ae4b0c2ea6b06a7461dbed8/src/main/java/trollogyadherent/offlineauth/gui/skin/SkinGuiRenderTicker.java#L316-L318) [2)](https://github.com/kotmatross28729/OfflineAuth/blob/1e48b6e8fc445ac28ae4b0c2ea6b06a7461dbed8/src/main/java/trollogyadherent/offlineauth/event/ClientEventListener.java#L327-L329) [3)](https://github.com/kotmatross28729/OfflineAuth/blob/1e48b6e8fc445ac28ae4b0c2ea6b06a7461dbed8/src/main/java/trollogyadherent/offlineauth/skin/client/ClientSkinUtil.java#L58-L60), but regarding the rest, you have no idea

**Comment by Roadhog360** [comment]
What the hell

**Comment by kotmatross28729** [comment]
So, the resolution issue (except for the 64 height) looks like a skin skill issue. And I think I figured out how it works.

The skins that work with SSB have the new format (2 layers and the ability to change 2nd leg/arm separately), and those that don't -  have the old format (and therefore don't have all of the above).

So this "bug" can be fixed by adding x2 height to the skin and copying the leg and arm textures to the correct positions.

As for the skins not working with 64 height: I'll fix it myself by simply checking if SSB is loaded, and if so, I'll simply disable all occurrences of LegacyConversion in the code.

EDIT: Fixed - [4d3d08c](https://github.com/kotmatross28729/OfflineAuth/commit/4d3d08cde7404b884385bdd35cac547f36ea460a)

**Comment by Roadhog360** [comment]
The first part is incorrect; this mod is supposed to work with 64x32 skins. If it isn't then that is a bug.

**Comment by Roadhog360** [comment]
Does OfflineAuth support slim arm data? If it doesn't I can try to find a way to make it work from my end later. If skins 100% work with your fork feel free to close as completed.

**Comment by kotmatross28729** [comment]
OfflineAuth itself does not support slim arm data, such skins will simply have black stripes on their arms.

SSB (with OfflineAuth ) forces the use of a slim arm format, this works on skins that have a slim arm format (obviously), but on regular skins, arms (and a part of the body) are shifted by 1 pixel

**Comment by Roadhog360** [comment]
Right, yeah. I wonder why, as SSB only activates slim arms when it receives that data from the player's game data, fetched from the auth endpoint. I'll have to see what OfflineAuth is doing. Maybe it is always reporting the arms as slim somehow

**Comment by Roadhog360** [comment]
I'm attempting to investigate OfflineAuth, would you be willing to open an OfflineAuth-enabled server so I can test fixes?

**Comment by kotmatross28729** [comment]
At the moment I don't have the ability to do this.

You can test on a local (dedicated LAN) forge server, or send the jar file to me, I know who can test it now

(Although I think that even an integrated server in a singleplayer is enough?)

**Comment by kotmatross28729** [comment]
So, I found a [slightly better way to have a new skin format with OfflineAuth](https://github.com/kotmatross28729/OfflineAuth/wiki/Guide:-New-skin-format-with-CPM): use Customizable Player Models.

I made the [necessary changes](https://github.com/kotmatross28729/OfflineAuth/commit/f256bf8627a4d32f07f283909501260e57765142) on the OfflineAuth side, and it seems to work.

Yeah, I know it's still an incompatibility with SSB, but I don't think it makes sense to fix something that can *already* work.

**Comment by Roadhog360** [comment]
Oh, okay. Thanks.
